mipengine/udfgen/udfgenerator.py

```diff
@@ -1,1043 +1,32 @@
-    `udf` decorator. Generator objects are instantiated based on a particular
-    function, found in a passed `UdfRegistry`, and on a set of arguments found
-    in the algorithm flow. Some additional flags and parameters are also
-    necessary. Then, the generator object has methods for generating two types
-    of queries, UDF definitions and UDF execution statements (basically a
-    SELECT wrapped in an INSERT statement), as well as objects representing the
-    resulting tables.
-    """
-
-    def __init__(
-        self,
-        udfregistry: UdfRegistry,
-        func_name: str,
-        flowargs: List[FlowUdfArg],
-        flowkwargs: Dict[str, FlowUdfArg],
-        smpc_used: bool = False,
-        request_id: Optional[str] = None,
-        output_schema=None,
-        min_row_count: int = None,
-    ):
-        """
-        Parameters
-        ----------
-        udfregistry : UdfRegistry
-            Data structure containing python functions registered as UDFs
-        func_name : str
-            Unique key of a function in `udfregistry`
-        flowargs : List[FlowUdfArg]
-            Positional arguments received by the algorithm flow
-        flowkwargs : Dict[str, FlowUdfArg]
-            Keyword arguments received by the algorithm flow
-        smpc_used : bool
-            True when SMPC framework is used
-        request_id : Optional[str]
-            Request id
-        output_schema : List[Tuple[str, DType]]
-            This argument needs to be provided when the output schema is declared as
-            deferred
-        min_row_count : int
-            Minimum allowed number of rows for an input table
-        """
-        self.func_name = func_name
-        self.smpc_used = smpc_used
-        self.request_id = request_id
-        self.output_schema = output_schema
-        self.min_row_count = min_row_count
-
-        self.funcparts = udfregistry[func_name]
-        if smpc_used:
-            cast_secure_transfers_for_smpc(self.funcparts.output_types)
-
-        self.udf_args = self._get_udf_args(flowargs, flowkwargs)
-
-    @functools.cached_property
-    def output_types(self) -> Tuple[OutputType, List[LoopbackOutputType]]:
-        """Computes the UDF output type.
-
-        There are three cases:
-            - The output type is `RelationType` whose `schema` is DEFERRED
-            - The output type is inferred dynamically
-            - The output type is known statically
-        """
 
-        input_types = copy_types_from_udfargs(self.udf_args)
+class UdfGenerator(ABC):
+    def __new__(cls, udfregistry, func_name, **kwargs):
+        if func_name in udfregistry:
+            return PyUdfGenerator(udfregistry, func_name, **kwargs)
+        elif AdhocUdfGenerator.is_registered(func_name):
+            return AdhocUdfGenerator.get_subclass(func_name)(**kwargs)
+        raise ValueError(f"UDF named '{func_name}' not found in registries.")
```

Comment:
Why don't the `AdhocUdfGenerator` and `PythondfGenerator` inherit from the `UdfGenerator` class and they are dynamically registered later?

Reply:
Because they would inherit `__new__` and go on an infinite loop.